scipy/sparse/linalg/_eigen/arpack/arpack.py

```diff
@@ -319,7 +319,12 @@ def __init__(self, n, k, tp, mode=1, sigma=None,
             raise ValueError("maxiter must be positive, maxiter=%d" % maxiter)
 
         if tp not in 'fdFD':
-            raise ValueError("matrix type must be 'f', 'd', 'F', or 'D'")
+            if np.can_cast(tp, 'f'):
+                tp = 'f'
+            elif np.can_cast(tp, 'd'):
+                tp = 'd'
```

Comment:
I would probably default to `float64` in the case of a non-floating-point input since that was the de-facto result of the old `dtype`-determination. Defaulting to `float32` caused a visible change in the precision of the outputs in the `svds()` docstring, which you had to adjust the `atol` for. That's undesirable.

Reply:
The  `svds()` docstring example is matrix-free, but both `matvec` and `rmatvec` there would correspond to multiplying by an integer-valued matrix resulting in the integer dtype of `aslinearoperator` consistently with the new behavior of the custom `LinearOperator`. 

The `svds()` docstring example calls ARPACK which before would give an error for integer types.  Now it works with integer types, calling the `float32` libraries. Maybe I should add the corresponding comments to the docstring example (which coincidently I am an author of...) explaining this rather than defaulting ARPACK to `float64` libraries for integer types while at the same time using `float32` libraries for `float32` types that is not quite logical? 

I do not think we need to be too much concerned here with backward compatibility, since such examples of `matvec` and `rmatvec` as used in this  `svds()` docstring example are very uncommon.

Reply:
When a function is implemented for both `float32` and `float64` and is given an integer input, we almost always use the `float64` codepath by default. `svds()` should do the same.

Reply:
Looks like I have already resolved this? @rkern